logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
+        findVariablesWithNoHints(conjunction);
 
-    private void addInferredSubLabels(TypeVariable variable, Set<Label> resolvedLabels) {
-        variable.addResolvedTypes(resolvedLabels);
+        return conjunction;
     }
 
-    private void addInferredIsaLabels(ThingVariable variable, Set<Label> resolveLabels) {
-        //TODO: use .getType(label) once ConceptManager can handle labels
-        resolveLabels.removeIf(label -> traversalEng.graph().schema().getType(label).isAbstract());
-        variable.addResolvedTypes(resolveLabels);
+    private void findVariablesWithNoHints(Conjunction conjunction) {
+        long numOfTypes = traversalEng.graph().schema().stats().thingTypeCount();
+        long numOfConcreteTypes = traversalEng.graph().schema().stats().concreteThingTypeCount();
+        iterate(conjunction.variables()).filter(variable -> (
+                variable.isType() && variable.resolvedTypes().size() == numOfTypes ||
+                        variable.isThing() && variable.resolvedTypes().size() == numOfConcreteTypes
+        )).forEachRemaining(Variable::clearResolvedTypes);
     }
 
-    private Map<Reference, Set<Label>> retrieveResolveTypes(Set<Variable> resolveVars) {
-        Conjunction resolveVariableConjunction = new Conjunction(resolveVars, Collections.emptySet());
-        resolveVariableConjunction = resolveLabels(resolveVariableConjunction);
-        return logicCache.resolver().get(resolveVariableConjunction, conjunction -> {
+    private Map<Reference, Set<Label>> runTraversalEngine(TraversalConstructor traversalConstructor) {
+        return logicCache.resolver().get(traversalConstructor.resolverTraversal(), traversal -> {
             Map<Reference, Set<Label>> mapping = new HashMap<>();
-            traversalEng.iterator(conjunction.traversal()).forEachRemaining(
+            traversalEng.iterator(traversal).forEachRemaining(
                     result -> result.forEach((ref, vertex) -> {
                         mapping.putIfAbsent(ref, new HashSet<>());
-                        mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
+                        assert vertex.isType();
+                        if (!(vertex.asType().isAbstract() && traversalConstructor.getVariable(ref).isThing()))
+                            mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
                     })
             );
             return mapping;
         });
     }
 
-    private Type getType(Label label) {
-        if (label.scope().isPresent()) {
-            assert conceptMgr.getRelationType(label.scope().get()) != null;
-            return conceptMgr.getRelationType(label.scope().get()).getRelates(label.name());
-        } else {
-            return conceptMgr.getThingType(label.name());
-        }
-    }
+    private static class TraversalConstructor {
```

Comment:
Essentially this class will take in a conjunction and create a `Traversal` that is the schematic representation of types induced by the conjunction.